<a href="https://colab.research.google.com/github/PLEX-GR00T/NLP/blob/main/Text_Processing%2C_Classification%2C_and_Topic_Modeling_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import nltk
import string
from nltk.corpus import movie_reviews
from nltk.corpus import stopwords

import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer
import numpy as np

In [3]:
nltk.download('movie_reviews')
nltk.download('stopwords')

[nltk_data] Downloading package movie_reviews to /root/nltk_data...
[nltk_data]   Package movie_reviews is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

# Pre-process

In [ ]:
reviews = [movie_reviews.raw(fileid) for fileid in movie_reviews.fileids()]
# reviews[0]

In [ ]:
# reviews[1]

In [8]:
# Remove Puncturations
punctuations = string.punctuation
words_wo_puncts = [x for x in reviews if x not in punctuations]

# Remove the Stopwords
eng_stopwords = stopwords.words('english')
words_wo_puncts_stopwords =[x for x in words_wo_puncts if x not in eng_stopwords] 

In [9]:
# Convert words to np array.
# movie1 = np.array(words_wo_puncts_stopwords)
# type(movie1)
movie2 = words_wo_puncts_stopwords

# Task 1: Count Vectors

In [10]:
# Only Unigrams
count_vect = CountVectorizer(analyzer='word', 
                             ngram_range=(1, 1), 
                             token_pattern = '[A-Za-z]+')
count_matrix = count_vect.fit_transform(movie2)
count_vect.get_feature_names_out()

array(['a', 'aa', 'aaa', ..., 'zwigoff', 'zycie', 'zzzzzzz'], dtype=object)

In [11]:
print("Only Unigram Shape: ",count_matrix.shape)

Only Unigram Shape:  (2000, 38911)


In [12]:
print(count_matrix.toarray())

[[14  0  0 ...  0  0  0]
 [13  0  0 ...  0  0  0]
 [10  0  0 ...  0  0  0]
 ...
 [31  0  0 ...  0  0  0]
 [ 7  0  0 ...  0  0  0]
 [20  0  0 ...  0  0  0]]


In [13]:
# Unigrams and Bigrams
count_vect = CountVectorizer(analyzer='word', 
                             ngram_range=(1, 2), 
                             token_pattern = '[A-Za-z]+')
count_matrix = count_vect.fit_transform(movie2)
count_vect.get_feature_names_out()

array(['a', 'a a', 'a abraded', ..., 'zycie za', 'zzzzzzz',
       'zzzzzzz critique'], dtype=object)

In [14]:
print("Unigram and Bigrams shape: ",count_matrix.shape)

Unigram and Bigrams shape:  (2000, 519635)


In [15]:
print(count_matrix.toarray())

[[14  0  0 ...  0  0  0]
 [13  0  0 ...  0  0  0]
 [10  0  0 ...  0  0  0]
 ...
 [31  0  0 ...  0  0  0]
 [ 7  0  0 ...  0  0  0]
 [20  0  0 ...  0  0  0]]


In [16]:
# Only Bigrams
count_bi_vect = CountVectorizer(analyzer='word', 
                             ngram_range=(2, 2), 
                             token_pattern = '[A-Za-z]+')
count_bi_matrix = count_bi_vect.fit_transform(movie2)
count_bi_vect.get_feature_names_out()

array(['a a', 'a abraded', 'a abundence', ..., 'zycie masterfully',
       'zycie za', 'zzzzzzz critique'], dtype=object)

In [17]:
print("Only bigram shape: ",count_bi_matrix.shape)

Only bigram shape:  (2000, 480724)


In [18]:
print(count_matrix.toarray())

[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]


# Task 2: TF-idf

In [19]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [20]:
# Unigram Tfidf
tfid_vect = TfidfVectorizer(analyzer='word', 
                             ngram_range=(1,1), 
                             token_pattern = '[A-Za-z]+')
tfid_count = tfid_vect.fit_transform(movie2)
tfid_vect.get_feature_names_out()

array(['a', 'aa', 'aaa', ..., 'zwigoff', 'zycie', 'zzzzzzz'], dtype=object)

In [21]:
print("Tfid vector unigram shape: ", tfid_count.shape)

Tfid vector unigram shape:  (2000, 38911)


In [ ]:
print(tfid_count.toarray())

In [22]:
# Unigram and bigram Tfidf
tfid_vect = TfidfVectorizer(analyzer='word', 
                             ngram_range=(1,2), 
                             token_pattern = '[A-Za-z]+')
tfid_count = tfid_vect.fit_transform(movie)
tfid_vect.get_feature_names_out()

array(['a', 'a a', 'a abraded', ..., 'zycie za', 'zzzzzzz',
       'zzzzzzz critique'], dtype=object)

In [23]:
print("Tfid vector unigram and bigram shape: ", tfid_count.shape)

Tfid vector unigram and bigram shape:  (2000, 519635)


In [ ]:
print(tfid_count.toarray())

In [24]:
# bigram Tfidf
tfid_bi_vect = TfidfVectorizer(analyzer='word', 
                             ngram_range=(2,2), 
                             token_pattern = '[A-Za-z]+')
tfid_bi_count = tfid_bi_vect.fit_transform(movie2)
tfid_bi_vect.get_feature_names_out()

array(['a a', 'a abraded', 'a abundence', ..., 'zycie masterfully',
       'zycie za', 'zzzzzzz critique'], dtype=object)

In [25]:
print("Tfid vector bigram shape: ", tfid_count.shape)

Tfid vector bigram shape:  (2000, 519635)


In [ ]:
print(tfid_bi_count.toarray())

# Task 3: Bayesian classification

In [ ]:
print(len(movie_reviews.fileids()))
print(movie_reviews.categories())

In [27]:
import random 

corpus = [(list(movie_reviews.words(fileid)), category)
             for category in movie_reviews.categories()
             for fileid in movie_reviews.fileids(category)]
random.seed(123)
random.shuffle(corpus)

## Test-Train Split

In [62]:
from sklearn.model_selection import train_test_split
train, test = train_test_split(corpus, test_size = 0.30, random_state=16)

In [63]:
## Distrubtion for Train and Test
from collections import Counter
print(Counter([label for words, label in train]))
print(Counter([label for words, label in test]))

Counter({'pos': 712, 'neg': 688})
Counter({'neg': 312, 'pos': 288})


In [64]:
X_train = [' '.join(words) for words, label in train]
X_test = [' '.join(words) for words, label in test]
y_train = [label for words, label in train]
y_test = [label for words, label in test]

## Part 1: Unigram comparison.

### 1.1 Count Vectors Unigram

In [116]:
## Count Vector Unigram
c_vector = CountVectorizer (analyzer = 'word',
                            ngram_range=(1,1), # Unigram
                            min_df = 10, # igore any frequencey of words below 10
                            token_pattern = r'[a-zA-Z]+')
X_train_cvu = c_vector.fit_transform(X_train) # fit train
X_test_cvu= c_vector.transform(X_test) # transform test

print(X_train_cvu.shape)
print(X_test_cvu.shape)

(1400, 6459)
(600, 6459)


In [117]:
from sklearn.naive_bayes import MultinomialNB
model_gnb = MultinomialNB()
model_gnb.fit(X_train_cvu.toarray(), y_train)

MultinomialNB()

In [118]:
# create predictions
y_pred = model_gnb.predict(X_test_cvu)

# find f-1 score
from sklearn.metrics import f1_score
score = f1_score(y_test, y_pred, average='micro')
print('F-1 score : {}'.format(np.round(score,4)))

F-1 score : 0.8133


### 1.2 TF-idf Unigram

In [83]:
## Tfidf Vectors Unigram
tfidf_vec = TfidfVectorizer(analyzer = 'word',
                            ngram_range=(1,1),
                            min_df = 10, 
                            token_pattern = r'[a-zA-Z]+')
X_train_tfidfu = tfidf_vec.fit_transform(X_train) # fit train
X_test_tfidfu = tfidf_vec.transform(X_test) # transform test

print(X_train_tfidfu.shape)
print(X_test_tfidfu.shape)

(1400, 10826)
(600, 10826)


In [114]:
# from sklearn.naive_bayes import MultinomialNB
model_m = MultinomialNB()
model_m.fit(X_train_tfidfu.toarray(), y_train)

MultinomialNB()

In [115]:
# create predictions
y_pred = model_m.predict(X_test_tfidfu)

# find f-1 score
from sklearn.metrics import f1_score
score = f1_score(y_test, y_pred, average='micro')
print('F-1 score : {}'.format(np.round(score,4)))

F-1 score : 0.8017


## Part 2: Bigram Comparison.

### 2.1 Count Vectors Bigrams

In [150]:
## Count Vector Unigram
c_vector_bi = CountVectorizer (analyzer = 'word',
                            ngram_range=(2,2), # Unigram
                            min_df = 10, # igore any frequencey of words below 10
                            token_pattern = r'[a-zA-Z]+')
X_train_cvb = c_vector_bi.fit_transform(X_train) # fit train
X_test_cvb= c_vector_bi.transform(X_test) # transform test

print(X_train_cvb.shape)
print(X_test_cvb.shape)

(1400, 10160)
(600, 10160)


In [151]:
from sklearn.naive_bayes import MultinomialNB
model_gnb = MultinomialNB()
model_gnb.fit(X_train_cvb.toarray(), y_train)

MultinomialNB()

In [152]:
# create predictions
y_pred = model_gnb.predict(X_test_cvb)

# find f-1 score
# from sklearn.metrics import f1_score
score = f1_score(y_test, y_pred, average='micro')
print('F-1 score : {}'.format(np.round(score,4)))

F-1 score : 0.8183


### 2.2 TF-idf Bigrams

In [138]:
## Tfidf Vectors Unigram
tfidf_vecb = TfidfVectorizer(analyzer = 'word',
                            ngram_range=(2,2),
                            min_df = 10, 
                            token_pattern = r'[a-zA-Z]+')
X_train_tfidfub = tfidf_vecb.fit_transform(X_train) # fit train
X_test_tfidfub = tfidf_vecb.transform(X_test) # transform test

print(X_train_tfidfub.shape)
print(X_test_tfidfub.shape)

(1400, 10160)
(600, 10160)


In [139]:
# from sklearn.naive_bayes import MultinomialNB
model_m = MultinomialNB()
model_m.fit(X_train_tfidfub.toarray(), y_train)

MultinomialNB()

In [142]:
# create predictions
y_pred = model_m.predict(X_test_tfidfub)

# find f-1 score
from sklearn.metrics import f1_score
score = f1_score(y_test, y_pred, average='micro')
print('F-1 score : {}'.format(np.round(score,4)))

F-1 score : 0.83


## Results F-1 Scores

\begin{array}{|c|c|} \hline
 & CountVecotr & TF-IDF \\ \hline
Unigram & 0.8133 & 0.8017 \\ \hline
Bigram & 0.8183 & 0.83 \\\hline
\end{array}

# Part 4: Use nltk.collocations functions to calculate the top 10 bigram collocations based on PMI.

In [167]:
from nltk.collocations import *
bigram_measures = nltk.collocations.BigramAssocMeasures()
finder = BigramCollocationFinder.from_words(movie_reviews.words())
finder.nbest(bigram_measures.pmi, 10)

[('\x05\x05', 'euuugh'),
 ('05425', 'keywords'),
 ('165', 'mph'),
 ('1812', 'overture'),
 ('1926', 'novella'),
 ('1hr', '40mins'),
 ('289', '460'),
 ('460', '939'),
 ('_angel', 'heart_'),
 ('_animal', 'house_')]

In [175]:
bigram_measures = nltk.collocations.BigramAssocMeasures()
finder = BigramCollocationFinder.from_words(movie_reviews.words())
finder.apply_freq_filter(3)
finder.nbest(bigram_measures.pmi, 10)


[('_october', 'sky_'),
 ('alanis', 'morissette'),
 ('arija', 'bareikis'),
 ('battleship', 'potemkin'),
 ('bokeem', 'woodbine'),
 ('edie', 'mcclurg'),
 ('gabriella', 'pescucci'),
 ('gedde', 'watanabe'),
 ('heath', 'ledger'),
 ('igor', 'stravinsky')]

# Task 5: Topic modeling exercise

In [ ]:
!pip install pyLDAvis 

In [177]:
from __future__ import print_function
import pyLDAvis

import pyLDAvis.sklearn
pyLDAvis.enable_notebook()
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.decomposition import LatentDirichletAllocation
# Load 20 newsgroups dataset
# First, the 20 newsgroups dataset available in sklearn is loaded. As always, the headers, footers and quotes are removed.

/usr/local/lib/python3.7/dist-packages/past/types/oldstr.py:5: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3,and in 3.9 it will stop working
  from collections import Iterable
/usr/local/lib/python3.7/dist-packages/past/builtins/misc.py:4: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3,and in 3.9 it will stop working
  from collections import Mapping


In [180]:
# Check the API for this to use the categories.
filters = ['sci.med', 'sci.space', 'talk.politics.guns']
newsgroups = fetch_20newsgroups(remove=('headers', 'footers', 'quotes'), categories = filters)
docs_raw = newsgroups.data
print(len(docs_raw))

1733


In [181]:
# Convert to document-term matrix
# Next, the raw documents are converted into document-term matrix, possibly as raw counts or in TF-IDF form.

tf_vectorizer = CountVectorizer(strip_accents = 'unicode',
                                stop_words = 'english',
                                lowercase = True,
                                token_pattern = r'\b[a-zA-Z]{3,}\b',
                                max_df = 0.5, 
                                min_df = 10)
dtm_tf = tf_vectorizer.fit_transform(docs_raw)
print(dtm_tf.shape)

(1733, 2589)


In [182]:
tfidf_vectorizer = TfidfVectorizer(**tf_vectorizer.get_params())
dtm_tfidf = tfidf_vectorizer.fit_transform(docs_raw)
print(dtm_tfidf.shape)

/usr/local/lib/python3.7/dist-packages/sklearn/feature_extraction/text.py:2032: UserWarning: Only (<class 'numpy.float64'>, <class 'numpy.float32'>, <class 'numpy.float16'>) 'dtype' should be used. <class 'numpy.int64'> 'dtype' will be converted to np.float64.
  UserWarning,


(1733, 2589)


In [184]:
# Fit Latent Dirichlet Allocation models - this step takes a while!

# for TF DTM
lda_tf = LatentDirichletAllocation(n_components=15, random_state=0)
lda_tf.fit(dtm_tf)

# for TFIDF DTM
lda_tfidf = LatentDirichletAllocation(n_components=15, random_state=0)
lda_tfidf.fit(dtm_tfidf)
LatentDirichletAllocation(batch_size=128, doc_topic_prior=None,
             evaluate_every=-1, learning_decay=0.7,
             learning_method='online', learning_offset=10.0,
             max_doc_update_iter=100, max_iter=10, mean_change_tol=0.001,
             n_jobs=1, n_components=15, perp_tol=0.1, random_state=0,
             topic_word_prior=None, total_samples=1000000.0, verbose=0)

LatentDirichletAllocation(learning_method='online', n_components=15, n_jobs=1,
                          random_state=0)

In [185]:
# Visualizing the models with pyLDAvis, using the term frequency (tf) vectors
pyLDAvis.sklearn.prepare(lda_tf, dtm_tf, tf_vectorizer)

/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.7/dist-packages/pyLDAvis/_prepare.py:247: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  by='saliency', ascending=False).head(R).drop('saliency', 1)


PreparedData(topic_coordinates=              x         y  topics  cluster       Freq
topic                                                
6      0.122582 -0.077558       1        1  12.392180
1     -0.023124  0.204133       2        1  10.370253
14    -0.215929 -0.081323       3        1   9.377513
4     -0.248786 -0.040542       4        1   9.190920
12    -0.194006 -0.039647       5        1   8.496218
2     -0.066800 -0.028729       6        1   7.873237
3     -0.002921  0.086416       7        1   7.553935
5      0.074986 -0.089442       8        1   7.003577
9      0.170089 -0.037943       9        1   6.439404
10     0.094155 -0.123028      10        1   5.012300
11     0.115384 -0.116063      11        1   4.562170
7      0.001778  0.234920      12        1   3.648552
0      0.084729  0.146403      13        1   3.182278
8      0.068517  0.066378      14        1   2.846191
13     0.019344 -0.103974      15        1   2.051272, topic_info=        Term        Freq       Total Category  logprob  loglift
2182   space  986.000000  986.000000  Default  30.0000  30.0000
1021     gun  610.000000  610.000000  Default  29.0000  29.0000
720      edu  518.000000  518.000000  Default  28.0000  28.0000
1677  people  738.000000  738.000000  Default  27.0000  27.0000
1513    nasa  389.000000  389.000000  Default  26.0000  26.0000
...      ...         ...         ...      ...      ...      ...
676      don   17.774044  625.090549  Topic15  -4.9092   0.3266
1323    like   17.452035  602.840029  Topic15  -4.9275   0.3445
339   center   12.825867  210.662604  Topic15  -5.2354   1.0879
1253    know   13.584073  466.474647  Topic15  -5.1780   0.3504
1267   large   12.016408  161.345436  Topic15  -5.3006   1.2894

[883 rows x 6 columns], token_table=      Topic      Freq       Term
term                            
0         2  0.920809    abiding
0        13  0.032886    abiding
10        7  0.204167   accident
10        8  0.136112   accident
10        9  0.612502   accident
...     ...       ...        ...
2585     13  0.074364  yesterday
2585     14  0.037182  yesterday
2586      2  0.247465       york
2586      3  0.160125       york
2586      6  0.582272       york

[3730 rows x 3 columns], R=30, lambda_step=0.01, plot_opts={'xlab': 'PC1', 'ylab': 'PC2'}, topic_order=[7, 2, 15, 5, 13, 3, 4, 6, 10, 11, 12, 8, 1, 9, 14])